src/DataTypes/DataTypesBinaryEncoding.cpp

```diff
@@ -0,0 +1,705 @@
+    }
+}
+
+template <typename T>
+void encodeEnumValues(const DataTypePtr & type, WriteBuffer & buf)
+{
+    const auto & enum_type = assert_cast<const DataTypeEnum<T> &>(*type);
+    const auto & values = enum_type.getValues();
+    writeVarUInt(values.size(), buf);
+    for (const auto & [name, value] : values)
+    {
+        writeStringBinary(name, buf);
+        writeBinaryLittleEndian(value, buf);
+    }
+}
+
+template <typename T>
+DataTypePtr decodeEnum(ReadBuffer & buf)
+{
+    typename DataTypeEnum<T>::Values values;
+    size_t size;
+    readVarUInt(size, buf);
+    for (size_t i = 0; i != size; ++i)
+    {
+        String name;
+        readStringBinary(name, buf);
+        T value;
+        readBinaryLittleEndian(value, buf);
+        values.emplace_back(name, value);
+    }
+
+    return std::make_shared<DataTypeEnum<T>>(values);
+}
+
+template <typename T>
+void encodeDecimal(const DataTypePtr & type, WriteBuffer & buf)
+{
+    const auto & decimal_type = assert_cast<const DataTypeDecimal<T> &>(*type);
+    /// Both precision and scale should be less than 76, so we can decode it in 1 byte.
+    writeBinary(UInt8(decimal_type.getPrecision()), buf);
+    writeBinary(UInt8(decimal_type.getScale()), buf);
+}
+
+template <typename T>
+DataTypePtr decodeDecimal(ReadBuffer & buf)
+{
+    UInt8 precision;
+    readBinary(precision, buf);
+    UInt8 scale;
+    readBinary(scale, buf);
+    return std::make_shared<DataTypeDecimal<T>>(precision, scale);
+}
+
+void encodeAggregateFunction(const String & function_name, const Array & parameters, const DataTypes & arguments_types, WriteBuffer & buf)
+{
+    writeStringBinary(function_name, buf);
+    writeVarUInt(parameters.size(), buf);
+    for (const auto & param : parameters)
+        encodeField(param, buf);
+    writeVarUInt(arguments_types.size(), buf);
+    for (const auto & argument_type : arguments_types)
+        encodeDataType(argument_type, buf);
+}
+
+std::tuple<AggregateFunctionPtr, Array, DataTypes> decodeAggregateFunction(ReadBuffer & buf)
+{
+    String function_name;
+    readStringBinary(function_name, buf);
+    size_t num_parameters;
+    readVarUInt(num_parameters, buf);
+    Array parameters;
+    parameters.reserve(num_parameters);
+    for (size_t i = 0; i != num_parameters; ++i)
+        parameters.push_back(decodeField(buf));
+    size_t num_arguments;
+    readVarUInt(num_arguments, buf);
+    DataTypes arguments_types;
+    arguments_types.reserve(num_arguments);
+    for (size_t i = 0; i != num_arguments; ++i)
+        arguments_types.push_back(decodeDataType(buf));
+    AggregateFunctionProperties properties;
+    auto action = NullsAction::EMPTY;
+    auto function = AggregateFunctionFactory::instance().get(function_name, action, arguments_types, parameters, properties);
+    return {function, parameters, arguments_types};
+}
+
+}
+
+void encodeDataType(const DataTypePtr & type, WriteBuffer & buf)
```

Comment:
I only skimmed the PR but I wondered why the binary datatype encoding/decoding is not implemented in the respective datatypes, e.g. why don't we have (for example) a function `encodeBinary(WriteBuffer & buf)` in `DataTypeString`, `DataTypeNumberBase`, etc.? That way, stuff that logically belongs together is also located together.

Reply:
EDIT: Oh, I guess this is because during decoding we don't know the data type before we decoded it, therefore we can't have the decoding code within any `DataType*` classes.

Reply:
Yeah, decoding part cannot be implemented within `DataType*` classes and I wanted to keep both encoding/decoding in one place in the code.